File: csharp/BCUR/BCUR/Xoshiro256.cs
using System.Buffers.Binary;
using System.Security.Cryptography;

namespace BlockchainCommons.BCUR;

/// <summary>
/// Xoshiro256** pseudo-random number generator, used internally for fountain codes.
/// Seeded from a SHA-256 hash of input bytes.
/// </summary>
internal sealed class Xoshiro256
{
    private ulong _s0, _s1, _s2, _s3;

    private Xoshiro256(ulong s0, ulong s1, ulong s2, ulong s3)
    {
        _s0 = s0;
        _s1 = s1;
        _s2 = s2;
        _s3 = s3;
    }

    /// <summary>
    /// Creates a Xoshiro256** from a 32-byte seed (as produced by SHA-256).
    /// The seed bytes are interpreted as 4 big-endian u64 values,
    /// matching the Rust implementation's byte reordering (BE read → LE transmute → LE read).
    /// </summary>
    internal static Xoshiro256 FromSeed(byte[] seed)
    {
        if (seed.Length != 32)
            throw new ArgumentException("seed must be 32 bytes", nameof(seed));

        // The Rust transform reads 4 BE u64s, transmutes to bytes (native LE),
        // then reads back as LE u64s — the net effect is simply reading BE u64s.
        var s0 = BinaryPrimitives.ReadUInt64BigEndian(seed.AsSpan(0, 8));
        var s1 = BinaryPrimitives.ReadUInt64BigEndian(seed.AsSpan(8, 8));
        var s2 = BinaryPrimitives.ReadUInt64BigEndian(seed.AsSpan(16, 8));
        var s3 = BinaryPrimitives.ReadUInt64BigEndian(seed.AsSpan(24, 8));

        return new Xoshiro256(s0, s1, s2, s3);
    }

    /// <summary>
    /// Creates a Xoshiro256** from a string seed (SHA-256 hashed).
    /// </summary>
    internal static Xoshiro256 FromString(string seed)
    {
        var hash = SHA256.HashData(System.Text.Encoding.UTF8.GetBytes(seed));
        return FromSeed(hash);
    }

    /// <summary>
    /// Creates a Xoshiro256** from arbitrary bytes (SHA-256 hashed).
    /// </summary>
    internal static Xoshiro256 FromBytes(ReadOnlySpan<byte> data)
    {
        var hash = SHA256.HashData(data);
        return FromSeed(hash);
    }

    /// <summary>
    /// Creates a Xoshiro256** seeded from the CRC32 checksum of the given bytes.
    /// </summary>
    internal static Xoshiro256 FromCrc(ReadOnlySpan<byte> data)
    {
        var crc = Crc32.Checksum(data);
        var crcBytes = new byte[4];
        BinaryPrimitives.WriteUInt32BigEndian(crcBytes, crc);
        return FromBytes(crcBytes);
    }

    private static ulong RotateLeft(ulong x, int k)
    {
        return (x << k) | (x >> (64 - k));
    }

    /// <summary>
    /// Returns the next random u64.
    /// </summary>
    internal ulong Next()
    {
        var result = RotateLeft(_s1 * 5, 7) * 9;

        var t = _s1 << 17;

        _s2 ^= _s0;
        _s3 ^= _s1;
        _s1 ^= _s2;
        _s0 ^= _s3;

        _s2 ^= t;

        _s3 = RotateLeft(_s3, 45);

        return result;
    }

    /// <summary>
    /// Returns a random double in [0, 1).
    /// </summary>
    internal double NextDouble()
    {
        return (double)Next() / ((double)ulong.MaxValue + 1.0);
    }

    /// <summary>
    /// Returns a random integer in [low, high].
    /// </summary>
    internal ulong NextInt(ulong low, ulong high)
    {
        return (ulong)(NextDouble() * (double)(high - low + 1)) + low;
    }

    /// <summary>
    /// Returns a random byte in [0, 255].
    /// </summary>
    internal byte NextByte()
    {
        return (byte)NextInt(0, 255);
    }

    /// <summary>
    /// Returns n random bytes.
    /// </summary>
    internal byte[] NextBytes(int n)
    {
        var result = new byte[n];
        for (int i = 0; i < n; i++)
        {
            result[i] = NextByte();
        }
        return result;
    }

    /// <summary>
    /// Returns a shuffled copy of the given list.
    /// </summary>
    internal List<T> Shuffled<T>(List<T> items)
    {
        var remaining = new List<T>(items);
        var shuffled = new List<T>(items.Count);
        while (remaining.Count > 0)
        {
            var index = (int)NextInt(0, (ulong)(remaining.Count - 1));
            shuffled.Add(remaining[index]);
            remaining.RemoveAt(index);
        }
        return shuffled;
    }

    /// <summary>
    /// Chooses a degree for fountain coding using weighted sampling.
    /// </summary>
    internal int ChooseDegree(int length)
    {
        var weights = new double[length];
        for (int i = 0; i < length; i++)
        {
            weights[i] = 1.0 / (i + 1);
        }
        var sampler = new WeightedSampler(weights);
        return sampler.Next(this) + 1;
    }
}

File: go/bcur/xoshiro256.go
package bcur

import (
	"crypto/sha256"
	"encoding/binary"
	"hash/crc32"
	"math/bits"
)

// xoshiro256 implements the Xoshiro256** PRNG algorithm.
type xoshiro256 struct {
	s [4]uint64
}

// newXoshiro256FromBytes creates a new Xoshiro256** PRNG seeded from arbitrary bytes.
// The bytes are hashed with SHA-256, then the 32-byte hash is interpreted as
// 4 big-endian uint64 values for the state.
func newXoshiro256FromBytes(data []byte) *xoshiro256 {
	hash := sha256.Sum256(data)
	return newXoshiro256FromHash(hash)
}

// newXoshiro256FromString creates a new Xoshiro256** PRNG seeded from a string.
func newXoshiro256FromString(s string) *xoshiro256 {
	return newXoshiro256FromBytes([]byte(s))
}

// newXoshiro256FromHash creates a new Xoshiro256** PRNG from a 32-byte SHA-256 hash.
func newXoshiro256FromHash(hash [32]byte) *xoshiro256 {
	x := &xoshiro256{}
	for i := 0; i < 4; i++ {
		x.s[i] = binary.BigEndian.Uint64(hash[8*i : 8*i+8])
	}
	return x
}

// newXoshiro256FromCRC creates a new Xoshiro256** PRNG seeded from the CRC32 of the given bytes.
func newXoshiro256FromCRC(data []byte) *xoshiro256 {
	checksum := crc32.ChecksumIEEE(data)
	var buf [4]byte
	binary.BigEndian.PutUint32(buf[:], checksum)
	return newXoshiro256FromBytes(buf[:])
}

// next returns the next uint64 from the PRNG.
func (x *xoshiro256) next() uint64 {
	result := bits.RotateLeft64(x.s[1]*5, 7) * 9

	t := x.s[1] << 17

	x.s[2] ^= x.s[0]
	x.s[3] ^= x.s[1]
	x.s[1] ^= x.s[2]
	x.s[0] ^= x.s[3]

	x.s[2] ^= t
	x.s[3] = bits.RotateLeft64(x.s[3], 45)

	return result
}

// nextDouble returns a uniformly distributed float64 in [0, 1).
func (x *xoshiro256) nextDouble() float64 {
	return float64(x.next()) / (float64(^uint64(0)) + 1.0)
}

// nextInt returns a uniformly distributed uint64 in [low, high].
func (x *xoshiro256) nextInt(low, high uint64) uint64 {
	return uint64(x.nextDouble()*float64(high-low+1)) + low
}

// nextByte returns a random byte (used for test utilities).
func (x *xoshiro256) nextByte() byte {
	return byte(x.nextInt(0, 255))
}

// nextBytes returns n random bytes (used for test utilities).
func (x *xoshiro256) nextBytes(n int) []byte {
	result := make([]byte, n)
	for i := range result {
		result[i] = x.nextByte()
	}
	return result
}

// shuffled returns a shuffled copy of items using Fisher-Yates.
func (x *xoshiro256) shuffled(items []int) []int {
	remaining := make([]int, len(items))
	copy(remaining, items)
	shuffled := make([]int, 0, len(items))
	for len(remaining) > 0 {
		index := int(x.nextInt(0, uint64(len(remaining)-1)))
		shuffled = append(shuffled, remaining[index])
		remaining = append(remaining[:index], remaining[index+1:]...)
	}
	return shuffled
}

// chooseDegree selects a degree for fountain encoding using harmonic weights.
func (x *xoshiro256) chooseDegree(length int) int {
	weights := make([]float64, length)
	for i := 0; i < length; i++ {
		weights[i] = 1.0 / float64(i+1)
	}
	sampler := newWeightedSampler(weights)
	return sampler.next(x) + 1
}

// makeMessage generates a deterministic pseudo-random byte sequence for test vectors.
func makeMessage(seed string, size int) []byte {
	rng := newXoshiro256FromString(seed)
	return rng.nextBytes(size)
}
